src/background.js

```diff
@@ -9,54 +9,101 @@ function log() {
   }
 }
 
+// Gate-keeping pref to run the add-on
+const DOH_ENABLED_PREF = "doh-rollout.enabled";
+
+// Pref that sets DoH to on/off. It has multiple modes:
+// 0: Off (default)
+// 1: null (No setting)
+// 2: Enabled, but will fall back to 0 on DNS lookup failure
+// 3: Always on.
+// 4: null (No setting)
+// 5: Never on.
 const TRR_MODE_PREF = "network.trr.mode";
 
+// This preference is set to TRUE when DoH has been enabled via the add-on. It will
+// allow the add-on to continue to function without the aid of the Normandy-triggered pref
+// of "doh-rollout.enabled". Note that instead of setting it to false, it is cleared.
+const DOH_SELF_ENABLED_PREF = "doh-rollout.self-enabled";
+
+// This pref is part of a cache mechnaism to see if the heuristics dictated a change in the DoH settings
+const DOH_PREVIOUS_TRR_MODE_PREF = "doh-rollout.previous.trr.mode";
+
+// Set after doorhanger has been interacted with by the user
+const DOH_DOORHANGER_SHOWN_PREF = "doh-rollout.doorhanger-shown";
+
+// Records if the user opted in/out of DoH study by clicking on doorhanger
+const DOH_DOORHANGER_USER_DECISION_PREF = "doh-rollout.doorhanger-decision";
+
+// Records if user has decided to opt out of study, either by disabling via the doorhanger,
+// unchecking "DNS-over-HTTPS" with about:preferences, or manually setting network.trr.mode
+const DOH_DISABLED_PREF = "doh-rollout.disable-heuristics";
+
 const stateManager = {
   async setState(state) {
     log("setState: ", state);
     browser.experiments.preferences.state.set({ value: state });
+
+    switch (state) {
+    case "uninstalled":
+      break;
+    case "disabled":
+      rollout.setSetting(TRR_MODE_PREF, 0);
+      browser.experiments.preferences.clearUserPref(DOH_SELF_ENABLED_PREF);
```

Comment:
I don't think we should do this here. Here, "disabled" means the add-on was disabled right? That might happen for a variety of reasons (like armagaddon or some other anomaly) and there's no need to forget this value - it would be useful if we get re-enabled.

Reply:
Great call. You're right!